pkg/virt-controller/watch/vm_test.go

```diff
@@ -673,6 +679,214 @@ var _ = Describe("VirtualMachine", func() {
+
+					}
+				}).Return(nil, nil)
+
+				//	if runStrategy != v1.RunStrategyManual {
+				shouldExpectVMIFinalizerRemoval(vmi)
+				//	}
+
+				controller.Execute()
+
+				if runStrategy != v1.RunStrategyManual {
+					testutils.ExpectEvent(recorder, SuccessfulDeleteVirtualMachineReason)
+				}
+			},
+
+				table.Entry("runStrategyHalted", v1.RunStrategyHalted),
+				table.Entry("always", v1.RunStrategyAlways),
+				table.Entry("manual", v1.RunStrategyManual),
+				table.Entry("rerunOnFailure", v1.RunStrategyRerunOnFailure),
+			)
+
+			table.DescribeTable("should calculated expected backoff delay", func(failCount, minExpectedDelay int, maxExpectedDelay int) {
+
+				for i := 0; i < 1000; i++ {
+					delay := calculateStartBackoffTime(failCount, defaultMaxCrashLoopBackoffDelaySeconds)
+
+					if delay > maxExpectedDelay {
+						Expect(fmt.Errorf("delay: %d: failCount %d should not result in a delay greater than %d", delay, failCount, maxExpectedDelay)).To(BeNil())
+					} else if delay < minExpectedDelay {
+						Expect(fmt.Errorf("delay: %d: failCount %d should not result in a delay less than than %d", delay, failCount, minExpectedDelay)).To(BeNil())
+
+					}
+				}
+			},
+
+				table.Entry("failCount 0", 0, 10, 15),
+				table.Entry("failCount 1", 1, 10, 15),
+				table.Entry("failCount 2", 2, 40, 60),
+				table.Entry("failCount 3", 3, 90, 135),
+				table.Entry("failCount 4", 4, 160, 240),
+				table.Entry("failCount 5", 5, 250, 300),
+				table.Entry("failCount 6", 6, 300, 300),
+			)
+
+			table.DescribeTable("has start failure backoff expired", func(vm *v1.VirtualMachine, expected int64) {
+				seconds := startFailureBackoffTimeLeft(vm)
+
+				if expected > 0 {
+					// since the tests all run in parallel, it's difficult to
+					// do precise timing. We set the `retryAfter` time but the test
+					// execution may happen seconds later. We use big numbers and
+					// account for some jitter to make sure the calculation falls within
+					// the ballpark of what we expect.
+					parallelTestJitter := expected / 10
+					if (expected - seconds) > parallelTestJitter {
+						Expect(seconds).To(Equal(expected))
+					}
+				}
+
+			},
+
+				table.Entry("no vm start failures",
+					&v1.VirtualMachine{},
+					int64(0)),
+				table.Entry("vm failure waiting 300 seconds",
+					&v1.VirtualMachine{
+						Status: v1.VirtualMachineStatus{
+							StartFailure: &v1.VirtualMachineStartFailure{
+								RetryAfterTimestamp: &metav1.Time{
+									Time: time.Now().Add(300 * time.Second),
+								},
+							},
+						},
+					},
+					int64(300)),
+				table.Entry("vm failure 300 seconds past retry time",
+					&v1.VirtualMachine{
+						Status: v1.VirtualMachineStatus{
+							StartFailure: &v1.VirtualMachineStartFailure{
+								RetryAfterTimestamp: &metav1.Time{
+									Time: time.Now().Add(-300 * time.Second),
+								},
+							},
+						},
+					},
+					int64(0)),
+			)
+		})
```

Comment:
I realize this is merged for quite some time, but resurrecting this following a unit tests failure I saw on the CI ([this](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6190/pull-kubevirt-goveralls/1423671111240912896)).
Looking closely at this test, I reckon that:
- In table entries 1 and 3, `expected` is always 0, so nothing is actually being asserted (`if expected > 0 { ... }`).
- In entry 2, if test execution happens >30 seconds after the evaluation of the entry, it's bound to fail, and otherwise silently pass without any assertion being made.

Actually, I'm not entirely sure what the purpose of this specific test. Also, given the logic of `calculateStartBackoffTime()` is quite simple and is anyway covered by the larger testcases in the `Context("crashloop backoff tests")`, do you think we could just remove it?

@davidvossel WDYT?

Reply:
#6203